### PR TITLE
feat(downloads): support downloads on cr and wk

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: src/browser.ts [hunk 1]
```diff
@@ -17,21 +17,31 @@
...
   close(): Promise<void>;
-  _setDebugFunction(debugFunction: (message: string) => void): void;
 }
 
-export async function createPageInNewContext(browser: Browser, options?: BrowserContextOptions): Promise<Page> {
-  const context = await browser.newContext(options);
-  const page = await context.newPage();
-  page._ownedContext = context;
-  return page;
+export abstract class BrowserBase extends EventEmitter implements Browser {
```
Comment (yury-s) [inline src/browser.ts:30]: nit: Browser already extends EventEmitter

File: src/chromium/crBrowser.ts [hunk 2]
```diff
@@ -275,12 +266,19 @@ export class CRBrowserContext extends BrowserContextBase {
   }
 
   async _initialize() {
+    const promises: Promise<any>[] = [
+      this._browser._session.send('Browser.setDownloadBehavior', {
```
Comment (yury-s) [inline src/chromium/crBrowser.ts:270]: Shouldn't contextId be passed too?

File: src/download.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+import * as path from 'path';
+import * as fs from 'fs';
+import * as util from 'util';
+import { Page } from './page';
+import { Events } from './events';
+import { Readable } from 'stream';
+
+export class Download {
+  private _downloadsPath: string;
```
Comment (yury-s) [inline src/download.ts:25]: nit: readonly?

File: docs/api.md [hunk 4]
```diff
@@ -2971,6 +2980,51 @@ const { chromium } = require('playwright');  // Or 'firefox' or 'webkit'.
...
+
+Returns readable stream for current download or `null` if download failed.
+
+#### download.delete()
+- returns: <[Promise]>
+
+Deletes the downloaded file.
+
+#### download.error()
+- returns: <[Promise]<null|[string]>>
```
Comment (JoelEinbinder) [inline docs/api.md:3019]: What kind of errors are returned here? When should I use this string?

File: src/server/processLauncher.ts [hunk 5]
```diff
@@ -23,8 +23,13 @@ import * as stream from 'stream';
 import * as util from 'util';
 import { TimeoutError } from '../errors';
 import { helper } from '../helper';
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
```
Comment (yury-s) [inline src/server/processLauncher.ts:28]: nit: run 'Organize imports'?

File: src/transport.ts [hunk 6]
```diff
@@ -184,3 +184,38 @@ export class SequenceNumberMixer<V> {
     return value;
   }
 }
+
+export class InterceptingTransport implements ConnectionTransport {
+  private readonly _delegate: ConnectionTransport;
+  private _interceptor: (message: ProtocolRequest) => ProtocolRequest;
```
Comment (yury-s) [inline src/transport.ts:190]: nit: readonly too?

File: src/webkit/wkPage.ts [hunk 7]
```diff
@@ -340,6 +340,10 @@ export class WKPage implements PageDelegate {
     this._page._frameManager.frameLifecycleEvent(frameId, event);
   }
 
+  private _onDownloadCreated() {
```
Comment (yury-s) [inline src/webkit/wkPage.ts:343]: revert?

File: test/download.spec.js [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+      ]);
+      const path1 = await download1.path();
+      const path2 = await download2.path();
+      expect(fs.existsSync(path1)).toBeTruthy();
+      expect(fs.existsSync(path2)).toBeTruthy();
+      await page.context().close();
+      expect(fs.existsSync(path1)).toBeFalsy();
+      expect(fs.existsSync(path2)).toBeFalsy();
+    });
+    it.fail(true)('should delete downloads on browser gone', async ({ server, defaultBrowserOptions }) => {
```
Comment (yury-s) [inline test/download.spec.js:99]: Unskip now that Joel has a fix?

**Review by yury-s — APPROVED** [review]

File: src/browserContext.ts [hunk 9]
```diff
@@ -89,13 +92,16 @@ export abstract class BrowserContextBase extends ExtendedEventEmitter implements
   _browserClosed() {
     for (const page of this.pages())
       page._didClose();
-    this._didCloseInternal();
+    this._didCloseInternal(true);
   }
 
-  _didCloseInternal() {
+  async _didCloseInternal(omitDeleteDownloads = false) {
```
Comment (JoelEinbinder) [inline src/browserContext.ts:98]: This was hard for me to follow. Can we leave it as `_didCloseInternal()`, and add `async deleteDownloads()`. Is it important that the close event fires before the downloads have been deleted?

File: src/chromium/crConnection.ts [hunk 10]
```diff
@@ -35,7 +38,6 @@ export class CRConnection extends EventEmitter {
   private readonly _sessions = new Map<string, CRSession>();
   readonly rootSession: CRSession;
   _closed = false;
-  _debugProtocol: debug.IDebugger;
```
Comment (JoelEinbinder) [inline src/chromium/crConnection.ts:38]: Is this intentional in this patch?

Comment (JoelEinbinder) [reply src/chromium/crConnection.ts:38]: Ah I see

File: utils/doclint/check_public_api/index.js [hunk 11]
```diff
@@ -205,7 +205,8 @@ function compareDocumentations(actual, expected) {
       [/Handle\<R\>/g, 'JSHandle'],
       [/JSHandle\<Object\>/g, 'JSHandle'],
       [/object/g, 'Object'],
-      [/Promise\<T\>/, 'Promise<Object>']
+      [/Promise\<T\>/, 'Promise<Object>'],
+      [/Readable/, 'stream.Readable']
```
Comment (JoelEinbinder) [inline utils/doclint/check_public_api/index.js:209]: If we leave this as `Readable`, it shouldn't need any special casing or overrides.

File: src/download.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    if (!this._acceptDownloads)
+      throw new Error('Pass { acceptDownloads: true } when you are creating your browser context.');
+    const fileName = path.join(this._downloadsPath, this._uuid);
+    await this._finishedPromise;
+    if (this._error)
+      return null;
+    return fileName;
+  }
+
+  async error(): Promise<string | null> {
```
Comment (yury-s) [inline src/download.ts:55]: May be `waitForFinished()` ?

**Review by yury-s — APPROVED** [review]

File: test/playwright.spec.js [hunk 13]
```diff
@@ -132,12 +132,12 @@ module.exports.describe = ({testRunner, product, playwrightPath}) => {
       state._stdout.on('line', dumpout);
       state._stderr.on('line', dumperr);
       if (dumpProtocolOnFailure)
-        state.browser._setDebugFunction(data => test.output.push(`\x1b[32m[pw:protocol]\x1b[0m ${data}`));
+        debug('pw:protocol').output = data => test.output.push(`\x1b[32m[pw:protocol]\x1b[0m ${data}`);
```
Comment (JoelEinbinder) [inline test/playwright.spec.js:135]: in my console: `debug('str') !== debug('str')`
How does this work?

File: docs/api.md (hunk 4 above)
Comment (pavelfeldman) [reply docs/api.md:3019]: WebKit will report network error here. If you did not pass `acceptDownloads`, you get `canceled` error as well.

File: src/browser.ts (hunk 1 above)
Comment (pavelfeldman) [reply src/browser.ts:30]: My understanding is that Browser is an interface, so it can't extend class

File: src/chromium/crBrowser.ts (hunk 2 above)
Comment (pavelfeldman) [reply src/chromium/crBrowser.ts:270]: Yes :)

File: src/download.ts (hunk 3 above)
Comment (pavelfeldman) [reply src/download.ts:25]: We don't do it systematically :/

File: src/server/processLauncher.ts (hunk 5 above)
Comment (pavelfeldman) [reply src/server/processLauncher.ts:28]: Sure

File: docs/api.md (hunk 4 above)
Comment (JoelEinbinder) [reply docs/api.md:3019]: Can this match the `response.failure()` api? It is equivalent right?

File: docs/api.md [hunk 14]
```diff
@@ -729,6 +733,11 @@ Emitted when a JavaScript dialog appears, such as `alert`, `prompt`, `confirm` o
 
 Emitted when the JavaScript [`DOMContentLoaded`](https://developer.mozilla.org/en-US/docs/Web/Events/DOMContentLoaded) event is dispatched.
 
+#### event: 'download'
+- <[Download]>
+
+Emitted when attachment is downloaded. User can access basic file operations on downloaded content via the passed [Download] instance. Browser context must be created with the `acceptDownloads` set to `true` when user needs access to the downloaded content. If `acceptDownloads` is not set or set to `false`, download events are emitted, but the actual download is not performed and user has no access to the downloaded files.
```
Comment (dgozman) [inline docs/api.md:739]: super nit: we don't usually mention "user" in the docs.

File: src/download.ts [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    this._finishedPromise = new Promise(f => this._finishedCallback = f);
+    this._page.emit(Events.Page.Download, this);
+    page._browserContext._downloads.add(this);
+    this._acceptDownloads = !!this._page._browserContext._options.acceptDownloads;
+  }
+
+  async path(): Promise<string | null> {
+    if (!this._acceptDownloads)
+      throw new Error('Pass { acceptDownloads: true } when you are creating your browser context.');
+    const fileName = path.join(this._downloadsPath, this._uuid);
```
Comment (dgozman) [inline src/download.ts:48]: Is there a way to retrieve suggested file name, extension, mime type or url? I imagine that would be very handy, e.g. to assert that my "Download" button tries to download the "screenshot-{current-time}.png" file.

File: src/download.ts [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+  async createReadStream(): Promise<Readable | null> {
+    const fileName = await this.path();
+    return fileName ? fs.createReadStream(fileName) : null;
+  }
+
+  async delete(): Promise<void> {
+    if (!this._acceptDownloads || this._deleted)
+      return;
+    const fileName = await this.path();
```
Comment (dgozman) [inline src/download.ts:70]: Should check `_deleted` after this await to avoid a race.

File: src/download.ts [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    this._deleted = true;
+    if (fileName && await util.promisify(fs.exists)(fileName))
```
Comment (dgozman) [inline src/download.ts:72]: Asynchronous `exists` check is racy. Better to always call `unlink` and catch `ENOENT` error.

File: test/download.spec.js [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        page.click('a')
+      ]);
+      const path1 = await download1.path();
+      const path2 = await download2.path();
+      expect(fs.existsSync(path1)).toBeTruthy();
+      expect(fs.existsSync(path2)).toBeTruthy();
+      await browser.close();
+      expect(fs.existsSync(path1)).toBeFalsy();
+      expect(fs.existsSync(path2)).toBeFalsy();
+      console.log(path.join(path1, '..'));
```
Comment (dgozman) [inline test/download.spec.js:120]: stray log

File: test/download.spec.js [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+      expect(fs.existsSync(path.join(path1, '..'))).toBeFalsy();
+    });
+    it('make coverage happy', async({page}) => {
+      // Otherwise FF gets mad.
```
Comment (dgozman) [inline test/download.spec.js:124]: There is a list of not covered methods in playwright.spec.js.

File: test/playwright.spec.js [hunk 20]
```diff
@@ -180,12 +180,12 @@ module.exports.addPlaywrightTests = ({testRunner, platform, products, playwright
...
           state._stderr.on('line', dumperr);
           if (dumpProtocolOnFailure)
-            state.browser._setDebugFunction(data => test.output.push(`\x1b[32m[pw:protocol]\x1b[0m ${data}`));
+            debug('pw:protocol').output = data => test.output.push(`\x1b[32m[pw:protocol]\x1b[0m ${data}`);
           state.tearDown = async () => {
             state._stdout.off('line', dumpout);
             state._stderr.off('line', dumperr);
             if (dumpProtocolOnFailure)
-              state.browser._setDebugFunction(() => void 0);
+              debug('pw:protocol').output = data => console.log;
```
Comment (dgozman) [inline test/playwright.spec.js:188]: Not sure what this meant to be, but looks suspicious :)

File: src/transport.ts [hunk 21]
```diff
@@ -184,3 +184,38 @@ export class SequenceNumberMixer<V> {
     return value;
   }
 }
+
+export class InterceptingTransport implements ConnectionTransport {
```
Comment (dgozman) [inline src/transport.ts:188]: Where is this one used?

File: docs/api.md (hunk 4 above)
Comment (pavelfeldman) [reply docs/api.md:3019]: It is not equivalent, but it is similar, so i can rename.

File: src/download.ts (hunk 12 above)
Comment (pavelfeldman) [reply src/download.ts:55]: Ugly?

File: src/download.ts (hunk 15 above)
Comment (pavelfeldman) [reply src/download.ts:48]: Yes, that'd be super handy. Follow up?

File: src/download.ts (hunk 16 above)
Comment (pavelfeldman) [reply src/download.ts:70]: Done!

File: src/download.ts (hunk 17 above)
Comment (pavelfeldman) [reply src/download.ts:72]: Done

File: test/download.spec.js (hunk 18 above)
Comment (pavelfeldman) [reply test/download.spec.js:120]: Done

File: test/download.spec.js (hunk 19 above)
Comment (pavelfeldman) [reply test/download.spec.js:124]: Done

File: test/playwright.spec.js (hunk 20 above)
Comment (pavelfeldman) [reply test/playwright.spec.js]: Fixed

File: src/transport.ts (hunk 21 above)
Comment (pavelfeldman) [reply src/transport.ts:189]: Nowhere just yet, but looks generally useful?

**Pull request merged**